Src/Extern/HYPRE/AMReX_HypreABecLap3.cpp

```diff
@@ -0,0 +1,503 @@
+}
+}
+
+namespace amrex {
+
+HypreABecLap3::HypreABecLap3(const BoxArray& grids,
+                             const DistributionMapping& dmap,
+                             const Geometry& geom_,
+                             MPI_Comm comm_)
+    : comm(comm_),
+      geom(geom_),
+      verbose(1),
+      A(NULL), b(NULL), x(NULL) {
+
+  const int ncomp = 1;
+  int ngrow = 0;
+  acoefs.define(grids, dmap, ncomp, ngrow);
+  acoefs.setVal(0.0);
+
+  for (int i = 0; i < BL_SPACEDIM; ++i) {
+    BoxArray edge_boxes(grids);
+    edge_boxes.surroundingNodes(i);
+    bcoefs[i].define(edge_boxes, dmap, ncomp, ngrow);
+    bcoefs[i].setVal(0.0);
+  }
+
+  bd.define(grids, dmap, ncomp, geom);
+
+  int num_procs, myid;
+  MPI_Comm_size(comm, &num_procs);
+  MPI_Comm_rank(comm, &myid);
+
+  // initialize with ngrow = 1
+  ngrow = 1;
+
+  // Store the global integer index
+  GbInd.define(grids, dmap, ncomp, ngrow);
+  GbInd.setVal(0);
+
+  // Arrays needed to build the global indices for all procs
+  CellsGIndex.resize(grids.size());
+  numCellsProc.resize(num_procs+1);
+
+  // These arrays store the starting global indices of all the boxes involved
+  StartIndex.resize(grids.size());
+
+  // Initialize all the arrays to 0
+  std::fill(CellsGIndex.begin(), CellsGIndex.end(), 0);
+  std::fill(numCellsProc.begin(), numCellsProc.end(), 0);
+  std::fill(StartIndex.begin(), StartIndex.end(), 0);
+
+  // Fill the numpoints in the box and in the proc where this box resides
+  for (int i = 0; i < grids.size(); i++) {
+    const Box& bx = grids[i];
+    CellsGIndex[i] = numCellsProc[ dmap[i]+1 ];
+    numCellsProc[ dmap[i]+1 ] = numCellsProc[ dmap[i]+1 ] + bx.numPts();
+  }
+
+  // making sure that counting starts from 0 for procId = 0
+  for (int i = 1; i< (num_procs+1); i++) {
+    if (numCellsProc[i] == 0) {
+      amrex::Abort("No rows in this core - Check domain decomposition");
+    }
+    numCellsProc[i] += numCellsProc[i-1];
+  }
+
+  // Box starting index starting from the proc offset
+  for (int i = 0; i < grids.size(); i++) {
+    CellsGIndex[i] = CellsGIndex[i] + numCellsProc[dmap[i]];
+  }
+
+  // Starting and ending global index for each box
+  for (int i = 0; i < (grids.size()-1); i++) {
+    StartIndex[i] = CellsGIndex[i];
+  }
+  StartIndex[grids.size()-1] = CellsGIndex[grids.size()-1];
+
+  // Fill up the imultifab with global indices
+  for (MFIter mfi(GbInd); mfi.isValid(); ++mfi) {
+    int i = mfi.index();
+    const Box &reg = mfi.validbox();
+
+    // build the global index for all cells on this level
+    amrex_BuildGlobalIndex(BL_TO_FORTRAN(GbInd[mfi]), ARLIM(reg.loVect()),
+                           ARLIM(reg.hiVect()), CellsGIndex[i]);
+  }
+
+  const int nghost = 0;
+  bool local = true;
+  int ilower = GbInd.min(0, nghost, local);
+  int iupper = GbInd.max(0, nghost, local);
+
+  // Fill the global indices in the ghost cells along the grid edges
+  GbInd.FillBoundary();
```

Comment:
Let's do `GbInd.FillBoundary(geom.periodicity())` to fill periodic boundary too.